apps/total-typescript/src/server/routers/pricing.ts

```diff
@@ -0,0 +1,143 @@
+import {createRouter} from 'server/createRouter'
+import {z} from 'zod'
+import {getSdk, Purchase} from '@skillrecordings/database'
+import {
+  formatPricesForProduct,
+  getActiveMerchantCoupon,
+} from '@skillrecordings/commerce-server'
+import {find} from 'lodash'
+import type {FormattedPrice} from '@skillrecordings/commerce-server/dist/@types'
+
+const merchantCouponSchema = z.object({
+  id: z.string(),
+  type: z.string(),
+})
+
+const PurchaseSchema: z.ZodType<Purchase> = z.any()
+
+const PricingFormattedInputSchema = z.object({
+  productId: z.string(),
+  userId: z.string().optional(),
+  quantity: z.number(),
+  couponId: z.string().optional(),
+  merchantCoupon: merchantCouponSchema.optional(),
+  upgradeFromPurchaseId: z.string().optional(),
+  code: z.string().optional(),
+  purchases: PurchaseSchema.array(),
+})
+
+const checkForAnyAvailableUpgrades = async ({
+  upgradeFromPurchaseId,
+  productId,
+  purchases,
+}: {
+  upgradeFromPurchaseId: string | undefined
+  productId: string
+  purchases: Purchase[]
+}) => {
+  if (upgradeFromPurchaseId) return upgradeFromPurchaseId
+
+  const {availableUpgradesForProduct} = getSdk()
+
+  const availableUpgrades = await availableUpgradesForProduct(
+    purchases,
+    productId,
+  )
+
+  return find(purchases, (purchase) => {
+    const upgradeProductIds = availableUpgrades.map(
+      (upgrade) => upgrade.upgradableFrom.id,
+    )
+    return upgradeProductIds.includes(purchase.productId)
+  })?.id
+}
+
+const CheckForAvailableCouponsSchema = PricingFormattedInputSchema.pick({
+  merchantCoupon: true,
+  couponId: true,
+  code: true,
+  productId: true,
+})
+type CheckForAvailableCoupons = z.infer<typeof CheckForAvailableCouponsSchema>
+
+const checkForAvailableCoupons = async ({
+  merchantCoupon,
+  couponId,
+  code,
+  productId,
+}: CheckForAvailableCoupons) => {
+  // explicit incoming merchant coupons are honored
+  // without checking for other potential coupons
+  // if there is no explicit incoming merchant coupon
+  // we check for default/global coupon or an incoming code
+  if (merchantCoupon?.id) {
+    return {
+      activeMerchantCoupon: merchantCoupon,
+      defaultCoupon: undefined,
+    }
+  } else {
+    const {activeMerchantCoupon, defaultCoupon} = await getActiveMerchantCoupon(
+      {
+        siteCouponId: couponId,
+        code,
+        productId,
+      },
+    )
+
+    const minimalDefaultCoupon = defaultCoupon && {
+      expires: defaultCoupon.expires?.toISOString(),
+      percentageDiscount: defaultCoupon.percentageDiscount.toString(),
+    }
+
+    return {activeMerchantCoupon, defaultCoupon: minimalDefaultCoupon}
+  }
+}
+
+export const pricing = createRouter().query('formatted', {
+  input: PricingFormattedInputSchema,
+  async resolve({ctx, input}): Promise<FormattedPrice> {
```

Comment:
Here is where the `pricing.formatted` trpc query lives -- I love this shift. This is a better spot for all of this logic and the built-in Zod input parsing is great.

Reply:
I'm imaging this query will get promoted to a skill-api query that can be shared with TA and other apps, but wanted to get the first pass working within just TT.

Reply:
Thinking about the `skill-api` I think it makes sense for webhooks/callbacks to some extent, but wonder how far we could take trpc and reduce the number of options we have for how we do things.

Reply:
the `purchases` option, for instance, is something that we could look up on the fly in a tRPC query instead of passing it all over. that would have the added benefit of ensuring the value is _fresh_.

Reply:
absolutely agree

Reply:
🎉  https://github.com/skillrecordings/products/pull/698